packages/react-sdk/src/components/BackgroundFilters/BackgroundFilters.tsx

```diff
@@ -197,166 +197,124 @@ export const BackgroundFiltersProvider = (
+  }, [backgroundFilter, call, start]);
 
-  const [isPlaying, setIsPlaying] = useState(false);
-  useEffect(() => {
-    if (!mediaStream || !videoRef) return;
-    const handleOnPlay = () => {
-      const [track] = mediaStream.getVideoTracks();
-      if (!track) return;
-      const { width: w = 0, height: h = 0 } = track.getSettings();
-      setWidth(w);
-      setHeight(h);
-      setIsPlaying(true);
-    };
-    videoRef.addEventListener('play', handleOnPlay);
-    videoRef.srcObject = mediaStream;
-    videoRef.play().catch((err) => {
-      console.error('Failed to play video', err);
-    });
-    return () => {
-      videoRef.removeEventListener('play', handleOnPlay);
-      videoRef.srcObject = null;
-      setIsPlaying(false);
-    };
-  }, [mediaStream, videoRef]);
+  return children;
+};
 
-  useEffect(() => {
-    const resolveFilter = signalFilterReadyRef.current;
-    if (!canvasRef || !resolveFilter) return;
+const useRenderer = (tfLite: TFLite) => {
+  const { backgroundFilter, backgroundBlurLevel, backgroundImage } =
+    useBackgroundFilters();
+  const videoRef = useRef<HTMLVideoElement>(null);
+  const canvasRef = useRef<HTMLCanvasElement>(null);
+  const bgImageRef = useRef<HTMLImageElement>(null);
+  const [videoSize, setVideoSize] = useState<{ width: number; height: number }>(
+    {
+      width: 1920,
+      height: 1080,
+    },
+  );
 
-    const filter = canvasRef.captureStream();
-    resolveFilter(filter);
-    return () => {
-      disposeOfMediaStream(filter);
-    };
-  }, [canvasRef]);
+  const start = useCallback(
+    (ms: MediaStream) => {
+      return new Promise<[MediaStream, () => void]>((resolve, reject) => {
+        if (!backgroundFilter) {
+          reject(new Error('No filter specified'));
+          return;
+        }
+
+        const videoEl = videoRef.current;
+        const canvasEl = canvasRef.current;
+        const bgImageEl = bgImageRef.current;
+
+        if (!videoEl || !canvasEl || (backgroundImage && !bgImageEl)) {
+          // You should start renderer in effect or event handlers
+          reject(new Error('Renderer started before elements are ready'));
+          return;
+        }
+
+        const [track] = ms.getVideoTracks();
+        const trackSettings = track.getSettings();
+        setVideoSize({
+          width: trackSettings.width ?? 0,
+          height: trackSettings.height ?? 0,
+        });
+        videoEl.srcObject = ms;
+        videoEl.play().then(() => {
+          const renderer = createRenderer(tfLite, videoEl, canvasEl, {
+            backgroundFilter,
+            backgroundBlurLevel,
+            backgroundImage: bgImageEl ?? undefined,
+          });
+
+          const outputStream = canvasEl.captureStream();
+
+          resolve([
+            outputStream,
+            () => {
+              renderer.dispose();
+              videoEl.srcObject = null;
+              disposeOfMediaStream(outputStream);
+            },
+          ]);
```

Comment:
With the new filter API, filter returns a stream and a cleanup function. The cleanup function will be called when the filter is unregistered or when the source stream is muted.